### DISCUSSION TITLE
Some questions

### DISCUSSION THREAD

**Discussion by nojaf** [Lanayx/Oxpecker#4] [General]
Hello, when reading your [tweet](https://x.com/Lanayx/status/1748988694717997225?s=20) I couldn't help but wonder about a few things here.

These are all rather meta aspects of this repository and there won't be anything actionable. So if you don't feel like responding to this, I respect that and feel free to close this at any point.

Here goes:

- I can't help but think that the original problem statement here is that the maintainer of Giraffe not being as active as people would like. This is a genuine concern, but I wonder if communication has really failed here or wasn't attempted enough. Did you get Dustin's perspective first-hand via a virtual meeting? Not that you need it, but do you have his blessing for this altered fork? I think it would help this story if that is the case. The way the blog post was written, I cannot shake the sense that the two of you never had a serious conversation about this. Maybe things could have been resolved without going the fork route. If you both didn't see eye-to-eye, that is fine and this outcome is still okay, but did that conversation even happen?

- How are you not going to fall for the same trap? One of the issues with Giraffe seems to be that only a single person can release new NuGet packages. The reason this repository seems to exist is to prevent such a state from happening, but why did you not solve this issue first and announce this project when there were three or more maintainers on board in this quest? Because right now there is a lot of work left to do to get on a similar level as Giraffe. Getting started guides, documentation, starter templates, contribution guides, etc... These all take a lot of work and when it is just you, how far will you get before you get burned out as well?

- Did you check anything with the SAFE stack before this move happened? This is again not something you are required to do, but as far as I understand this, they also have a stake in the health of Giraffe. My point is, did you really check with enough sources in the community beforehand? You could have found some co-maintainers first before diving into the code.

- `and move it to fsprojects once the project gets 200 stars`, not a question here but some feedback: 200 stars is a poor metric to decide if something should go to fsprojects. What you are looking for is people you can trust to work with you on this. People you can give the publish keys to and can operate the project without your intervention.

To summarize my thoughts on this, I fear that good communication could have been lacking in this story. Did all parties really hear each other? Does everyone understand why there is friction?

The F# community is a small one, most popular projects only have one maintainer and that is problematic.
People do not talk enough with each other and I cannot help but feel that all of this is the direct result of a human problem.

**Comment by Lanayx** [comment]
Dear @nojaf , Thanks for the issues raised, let me try to address them

- I was eager to join Giraffe and help evolve the project and tried to reach Dustin but was unable to do that. He actually ignored all the conversation attempts about Giraffe. Also there is a series of facts that justified the fork for me: right now Giraffe org has several people, but nobody can merge PR or publish nuget package without Dustin, there is no maintenance happening for the last half of a year, and before that there was a long period of unhealthy inactivity in the repo as well, also one of the org members (@TheAngryByrd) clearly stated that he was ignored as well. 
- I'm an experienced maintainer, so I don't need three or more maintainers for start, however I'm planning to add other people as well, if they are ready to contribute on regular basis. For example, I can add @TheAngryByrd to project admins right away, as a former Giraffe maintainer, if he asks me about that. The documentation piece and starters guides are the issue, but I hope that people will be ready to create content and contribute once they start using it, especially those who are already into Giraffe, since they are so similar. Still, my strong opinion is that proper maintenance is more important than proper documentation.
- As for the SAFE stack, they are already in trouble with Giraffe, so they will benefit from having community owned library. As for co-maintainers on pre-start, I believe that this works the other way, people would rather join a vivid project rather than something unknown and I'm open for collaboration.
- The stars although a poor metric, give general guidance - is the project in demand or not, it's also an easy to observe metric for someone who wants to understand how soon the project will be moved to fsprojects for shared responsibility. Again, I already have an experience with such move and I don't see any obstacles to do it again when the goal is achieved.

As a side note - the fork haven't happened out of the thin air. I'm starting a new project at work, so I didn't want to spend months at finding a way to approach Dustin and more months to agree on solution, and life is short! Also, there are particular things that I wanted to fix in Giraffe from it's inception, but which will be a breaking change for it, so a fork definitely gives me such a chance.

**Reply by TheAngryByrd** [reply]
I appreciate the offer but I was trying to go thru similar steps of transition to a maintainer that could commit more fully but I was never able to get there. I'm glad this kind of transition is finally happening. If you just want a backup "still has a key just in case" kind of maintainer, I can do that.

**Comment by isaacabraham** [comment]
From a SAFE Stack point of view (and the wider F# community), I think that there is value in trying to keep Giraffe going as a brand - the same as Saturn - because it's a known thing. Forking will unfortunately just create more confusion and split the community, so personally I would rather have that as a last resort. Moving to Giraffe / Saturn from Suave caused enough headaches and I really don't want to repeat anything like that again.

We really aren't in trouble with Giraffe, I chat with Dustin regularly, but simultaneously I think he would probably appreciate having some extra maintainers / contributors to the project.

**Comment by dustinmoris** [comment]
Hey, I actually only saw now your original tweet and blog post and had a read. I admit I have no interest in maintaining Giraffe any longer. Not just Giraffe but really any open source project. I just don't have the time for this. I still love programming and I work on a lot of side projects in my spare time, but nothing which is open source or visible to the public. Open source work is a lot of work where you have to constantly lend your brain capacity to the desires, questions and wishes of random people around the world for very little to almost no financial reward. It's something I don't enjoy doing anymore. I rather work on my own issues or on products which people pay money for as I find this line of work more satisfying whilst also being more in control of where and how I shape the software. This is basically a long explanation where I am mentally and why I've let Giraffe "rot". That is my mistake and when people reached out to me to help I often put it on the backburner thinking I need to take some time to reply to the kind people who want to help but then I never found the time to respond, which meant that I've actually done the opposite of what I wanted to do: I appeared to be ignoring people who reach out when in fact I just delayed it because I waited to devote them more attention than I was currently able to offer.

So long story short, if you want to can take full ownership of Giraffe. I can transfer everything to fsprojects today or make you the most powerful Admin of the giraffe-fsharp GitHub org or do it via another way if you have a preference. If however you want to start something new now then I also understand that and you have my full blessing! The F# community deserves a good functional web framework and I've not been able to live up to that over the last few years.

**Reply by Lanayx** [reply]
Thanks, Dustin! If you had written that a month ago and I wouldn't have done the Oxpecker work I would just be happy to take the ownership and continue with maintaining Giraffe. Now I see things a bit differently, so am not sure what is the best way to proceed. As I responded to Don in another message, my vision that a breaking change should happen with a series of improvements and I'm not sure where is the best place to make it happen. If you want to step out of giraffe right now, you can give me the admin rights of the giraffe-fsharp organization and nuget packages and I'll pass them to another person(s) who will be ready to maintain it (it might even be me if this is how things will go). Or you can wait until community decides what is best for Giraffe and then proceed with that solution.

**Reply by dustinmoris** [reply]
That's fair enough, I completely understand if you want to rather focus on Oxpecker now. All I can say is sorry that I neglected Giraffe. I can do as much as finding new maintainers myself. I agreed with Dom to give it a bit of time so more people have a chance to chime in and then I'll select two new core maintainers and let them select any additional ones as they see fit and do a proper transfer.

**Reply by fpellet** [reply]
Thank you for everything.
It's easy to flog yourself and that can lead to burnout.
It's great to explain the situation, admit your mistakes (nobody's perfect) and suggest solutions. 👍

I think the solution is to have more maintainers. I'm clearly not available to be the main maintainer on Giraffe, but I can participate.

**Reply by dustinmoris** [reply]
Anyone who is keen to take on Giraffe is welcome to throw in their hat! I've noted you down if you are interested.

**Reply by l3m** [reply]
@dustinmoris As we are using Giraffe in production and have a team of experienced devs working with Giraffe daily, I or we would be happy to help out.

**Reply by dsyme** [reply]
@l3m Thank you, that is great to hear!   Would that be at the level of being a project maintainer?

**Reply by l3m** [reply]
@dsyme I would think so. Ideally, Giraffe should not have a single maintainer, so I would suggest a chat in the fsharp slack or so for a start with interested parties.

**Reply by dustinmoris** [reply]
@fpellet @l3m If you two are happy to work together as new core maintainers on Giraffe then shall we find a day next week or the week after to hop on a call and I'll make you admins of giraffe-fsharp, ensure you have the NuGet keys to publish packages and discuss/answer any questions you have?

**Reply by dustinmoris** [reply]
If yes, let's continue this conversation outside of this issue. You can reach me on dustin@dusted.codes via email. Drop me a reply with either you're in or out and then I'll take it from there!

**Reply by fpellet** [reply]
Thanks for the suggestion, but I haven't put myself forward as a core maintainers because I don't have enough time.

If I'm not mistaken, @Lanayx and @64J0 have put themselves forward.

**Reply by dustinmoris** [reply]
Oh I misunderstood that. My understanding was that @Lanayx wanted to focus on Oxpecker now or have I got this wrong as well?

**Reply by Lanayx** [reply]
Dustin, you are right. My interest now is in building a new framework that will be competitive wth other modern frameworks by improving Giraffe codebase as much as I can and not looking at backward compatibility with Giraffe api. Keeping compatibility and supporting existing Giraffe codebases is not something I want to work on, so I can't lead Giraffe project itself.

**Reply by 64J0** [reply]
> If yes, let's continue this conversation outside of this issue. You can reach me on [dustin@dusted.codes](mailto:dustin@dusted.codes) via email. Drop me a reply with either you're in or out and then I'll take it from there!

Should I send the e-mail as well? I'm in btw.

**Reply by dustinmoris** [reply]
@64J0 Yes please if you send me an email so I got your email address and we find a slot either this or next week, that'd be great. I'm also following up with a few other people who showed interest because I think having 2 core maintainers would be preferable.

**Reply by 64J0** [reply]
Sent right now 🤝

**Comment by isaacabraham** [comment]
My strong, strong preference would be to keep the Giraffe brand going and make additions / breaking changes to that rather than a brand new project. I've looked through the blog post and there are some ideas in there that Dustin and I have already been chucking around in the past few months so I think we're all thinking along the same lines.

**Reply by Lanayx** [reply]
Thank you for the opinion! As I said in other messages, I really feel the need in core breaking changes, and I'm not sure if keeping just the same Giraffe repo and package will be user friendly, since they will look for a way to "upgrade" one version to another and will still remember old maintenance problems. And they will also complain about breaking changes. It reminds me the story with Angular and Angular2, where second version was really a brand new project, so it was no different - to switch to Angular2 or Aurelia. As for new users, it will be no difference for them to use Giraffe, Oxpecker or anything else, they will go to the place where documentation is better and there are enough stars and maintenance, also if someone or something will guide them, say a big message in giraffe's readme - you better go to another repository.

**Comment by dsyme** [comment]
Hi all, 

Cross linking this to keep the discussion in sync: https://github.com/fsprojects/.github/issues/49#issuecomment-1904259515

Whether "fsprojects" or "giraffe-fsharp" is used as the containing org, the main first thing is to identify 2 possible initial active maintainers. These are usually people who have recently contributed PRs and are motiviated to work on the project as a whole, including making releases.  

If the org "giraffe-fsharp" is used, then there should also be 1 or 2 backup maintainers who have ownership rights on org and nuget packages.

The fsprojects policies may be helpful here: https://github.com/fsprojects/.github/blob/master/profile/README.md

Don

**Comment by dsyme** [comment]
@Lanayx If you were to become a primary maintainer on Giraffe, I think it would be feasbile for you to do a Giraffe vNext (with breaking changes) alongside sustaining the current Giraffe.  That's effectively what happened to Fabulous, which is a similar framework-sized thing.

What do you think?  It feels like there could be a solution along these lines, maybe with @TheAngryByrd as initial co-maintainer?

I understand it's a different prospect for you though.

**Reply by dustinmoris** [reply]
Also circling in these users who were some of the most recent Giraffe contributors:

@TheAngryByrd
@64J0
@jcmrva
@kerams
@onpikono
@SIRHAMY

**Reply by dsyme** [reply]
It's possible this should be a discussion in a fresh issue in Giraffe repo, so we can bring in all interested until a way forward is found?

Should we create one there and link across?  Partly because solving the Giraffe maintenance problem may or may not align with Lanayx/Oxpecker, depending on how things sit.

**Reply by 64J0** [reply]
I'd like to become a co-maintainer too if possible.

**Reply by Lanayx** [reply]
@dsyme  Actually, my goal is to have a healthy web framework for F#. If this it to be Giraffe vNext with other maintainers, I'm ok :) The only regret for me would be 4 weeks of intensive work, but I can live with it. The main ideas brought up in Oxpecker I described in the blog post, hopefully at least some of them will make their way through.

**Reply by dsyme** [reply]
> @dsyme Actually, my goal is to have a healthy web framework for F#. ...The only regret for me would be 4 weeks of intensive work, but I can live with it.

:) Both approaches are absolutely valid - creating afresh is a great thing to do and totally part of community vibrancy.  Sometimes that needs to be done.  We've always had multiple web frameworks and regardless of particular decisions I'd encourage that to continue, it's been a marvelous part of the F# ecosystem.

It does sound like we have the makings of a way to take Giraffe forward here.  Let's give time for more people to chime in (it's morning in USA right now) and start to shape a plan over the next few days.

**Reply by Lanayx** [reply]
My opinion that is has to be a new thing. Giraffe's current endpoint routing was already a new thing and was really a breaking change, kinda repo inside repo. So a new project should be either Giraffe2 or GiraffeVNext or similar. It just can't stay the same Giraffe anymore. I think that even going forward with Oxpecker would be a good move, the brand value that @isaacabraham is talking about is a overestimated, it has lost a lot of it's value during the years of poor maintenance. For example the query "fsharp web framework" doesn't return it on the first page in Google, only the projects lists (which can be easily updated)
As for the maintenance of the new project, I can be both maintainer (as you can see with Oxpecker) or step away depending on the situation, but I think that having BDFL for such project would be a good thing. If talking about maintenance of the "old" Giraffe, then I'm not interested in that.

**Reply by fpellet** [reply]
I don't think that the work done on Oxpecker will be lost in any case.
I'm in favour of keeping Giraffe and proposing 2 ways of doing things to make migration easier, like what had started to be done with endpoints.
Starting again with another project will divide the community. I know too many projects that have done this and either died or took 10 years to merge the community again. Migration is important and having 2 completely different trunks isn't going to make it any easier.

After that, we're volunteers, you're free to do what you want, and in any case I'd like to thank everyone for their investment.
Giraffe remains a superb project :)

I'm clearly not available to be the main maintainer on Giraffe, but I can participate.

**Reply by dsyme** [reply]
@64J0 writes:

> I'd like to become a co-maintainer too if possible.

That is great!  We'll collect all the replies and start to decide the way forward, but I'm pretty confident we can have you as one of the co-maintainers.

**Reply by otto-gebb** [reply]
@Lanayx As far as breaking changes are concerned, Rich Hickey would [say](https://youtu.be/oyLBGkS5ICk?si=61gACst8rfxPs8ya&t=1937) "you might as well change the name", but going the Giraffe vNext route certainly has advantages:
- less community fragmentation;
- easier decision to upgrade for users: 2.0 explicitly supersedes 1.x and thereby communicates that it is supposed to be better and recommended;
- less work on docs, most of them remain applicable.

**Reply by Lanayx** [reply]
@otto-gebb De-facto, the decision has already been made :) Giraffe will be supported as is and Oxpecker will continue it's own life. As for community fragmentation - I would imagine that Giraffe should specialize on original implementation and Oxpecker should specialize on endpoint routing, so users would have a clear understanding of what to choose.

**Comment by i-shm** [comment]
"I'm clearly not available to be the main maintainer on Giraffe, but I can participate."

Likewise, I'm willing to help in any way I can (documentation, translations, etc.) to push F# forward!

**Reply by dustinmoris** [reply]
Everyone is welcome to throw in their hat. My plan (with the help from Don Syme and the F# Foundation) is to find at least two *core* maintainers who take ownership of Giraffe and facilitate the continuation and maintenance of the project. Then these two core maintainers can add more maintainers to the organisation as they see fit. The only requirement I have is that the people who volunteer to step in must be serious in wanting to maintain the project as this is the main issue we are trying to fix here. It will require more work than just updating the documentation once in a while. So long as people who throw in their hat are realistic about this and willing/keen to take over then you are definitely a candidate. I am basically waiting for more people who are keen to step up so we can pick two new maintainers.

**Comment by psantoro** [comment]
Outside of reporting bugs and supplying a work-around or two, I’ve never contributed to an open source project before.  However, I’d be willing to help out this project on a part-time basis, as I enjoy using F#.  Having experienced breaking changes many times (e.g. Microsoft’s ASP.NET 6.x release broke one of my F# Giraffe applications, ongoing churn with various client-side Javascript frameworks), here’s my 2cents on how to do this transition:

- keep the Giraffe project name (v2 or vNext is fine to differentiate the old/new releases)
- at first glance, I’m OK with the changes/assumptions made with the Oxpecker release
	- testing a real application migration may change my mind though
 		- perhaps this is how I can initially contribute to this project?
- provide sufficient migration documentation
	- should be tested on at least one real production-grade project
- provide deprecation warnings if/where possible

**Reply by dsyme** [reply]
@psantoro This is really helpful, thank you!
